src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs

```diff
@@ -1307,6 +1307,217 @@ static void Main()
+            CompileAndVerify(compilation, expectedOutput: @"2, 1+  1
+4, explicit-value").VerifyDiagnostics();
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void TestDelegate()
+        {
+            string source = @"
+using System;
+using System.Runtime.CompilerServices;
+
+class Program
+{
+    const string s1 = nameof(s1);
+    delegate void D(string s1, ref string s2, out string s3, [CallerArgumentExpression(s1)] string s4 = """");
+
+    static void M(string s1, ref string s2, out string s3, string s4 = """")
+    {
+        s3 = """";
+        Console.WriteLine($""s1: {s1}"");
+        Console.WriteLine($""s2: {s2}"");
+        Console.WriteLine($""s3: {s3}"");
+        Console.WriteLine($""s4: {s4}"");
+    }
+
+    public static void Main()
+    {
+        D d = M;
+        string s2 = ""s2-arg"";
+        d.Invoke(""s1-arg"", ref s2, out _);
+        //d.EndInvoke(ref s2, out _, null);
+    }
+}
+";
+
+            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe, parseOptions: TestOptions.RegularPreview);
+            CompileAndVerify(compilation, expectedOutput: @"s1: s1-arg
+s2: s2-arg
+s3: 
+s4: ""s1-arg""").VerifyDiagnostics();
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void TestDelegate2()
+        {
+            string source = @"
+using System;
+using System.Runtime.CompilerServices;
+
+class Program
+{
+    const string s3 = nameof(s3);
+    delegate void D(string s1, ref string s2, out string s3, [CallerArgumentExpression(s3)] string s4 = """");
+
+    static void M(string s1, ref string s2, out string s3, string s4 = """")
+    {
+        s3 = """";
+        Console.WriteLine($""s1: {s1}"");
+        Console.WriteLine($""s2: {s2}"");
+        Console.WriteLine($""s3: {s3}"");
+        Console.WriteLine($""s4: {s4}"");
+    }
+
+    public static void Main()
+    {
+        D d = M;
+        string s2 = ""s2-arg"";
+        d.Invoke(""s1-arg"", ref s2, out _);
+        //d.EndInvoke(ref s2, out _, null);
+    }
+}
+";
+
+            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe, parseOptions: TestOptions.RegularPreview);
+            CompileAndVerify(compilation, expectedOutput: @"s1: s1-arg
+s2: s2-arg
+s3: 
+s4: _").VerifyDiagnostics(); // PROTOTYPE(caller-arg): Should 'out' keyword be included?
+        }
+
+
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void TestDelegate3()
+        {
+            string source = @"
+using System;
+using System.Runtime.CompilerServices;
+
+class Program
+{
+    const string s1 = nameof(s1);
+    delegate void D(string s1, ref string s2, out string s3, [CallerArgumentExpression(s1)] string s4 = """");
```

Comment:
>[CallerArgumentExpression(s1)] [](http://example.com/codeflow?start=61&length=30)

I think the other interesting scenario is when the attribute is on parameter that is getting cloned for EndInvoke and it is pointing to the parameter that isn't getting cloned. Even though this is probably an error scenario, the symbol is going to be created and we need to check what does the API says for it. We also should make sure an attempt to consume the method isn't going to cause a failure in compiler

Reply:
Or the attribute is pointing to the parameter that is getting cloned, but that clone is placed is at the different ordinal by comparison to the original parameter.

Reply:
For both BeginInvoke and EndInvoke, all the parameters are guaranteed to be non-optional.

For C#, this guarantee comes from:

https://github.com/dotnet/roslyn/blob/22a72038d011eadbaa5039092fbab5f8a4d30650/src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateMethodSymbol.cs#L406

https://github.com/dotnet/roslyn/blob/22a72038d011eadbaa5039092fbab5f8a4d30650/src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateMethodSymbol.cs#L357

So, similar to VB, `CallerArgumentExpressionIndex` will be unreachable.

When I refactor `SourceClonedParameterSymbol`, I'll have `throw ExceptionUtilities.Unreachable`, in which case the existing tests seem enough to me. But I'm open to adding more tests if (after the refactoring) you think those tests are needed.